tests/tas-elasticsearch/src/test/java/org/alfresco/elasticsearch/ElasticsearchLimitTests.java

```diff
@@ -0,0 +1,119 @@
+    private static final int TOTAL_DOCUMENT_COUNT = 20;
+
+    @Autowired
+    private DataUser dataUser;
+
+    @Autowired
+    private DataContent dataContent;
+
+    @Autowired
+    private DataSite dataSite;
+
+    @Autowired
+    private ServerHealth serverHealth;
+
+    @Autowired
+    protected SearchQueryService searchQueryService;
+
+    private UserModel userSite1;
+    private SiteModel siteModel1;
+
+    @BeforeClass(alwaysRun = true)
+    public void dataPreparation()
+    {
+        serverHealth.assertServerIsOnline();
+
+        userSite1 = dataUser.createRandomTestUser();
+
+        siteModel1 = dataSite.usingUser(userSite1).createPrivateRandomSite();
+
+        dataUser.addUserToSite(userSite1, siteModel1, UserRole.SiteContributor);
+
+        for (int i = 0; i < TOTAL_DOCUMENT_COUNT; i++)
+        {
+            createContent(PREFIX + i + ".txt", "Document " + i, siteModel1, userSite1);
+        }
+    }
+
+    @TestRail(section = TestGroup.SEARCH, executionType = ExecutionType.REGRESSION, description = "Total hit tracking unset should use default behaviour")
+    @Test(groups = TestGroup.SEARCH)
+    public void searchUsingTotalHitsLimitDefaultValue()
+    {
+        SearchRequest request = req("cm:name:" + PREFIX + "*.txt");
+
+        // Without setting limits, we should get all up to 10k
+        searchQueryService.expectTotalHitsFromQuery(request, userSite1, TOTAL_DOCUMENT_COUNT);
+    }
+
+    @TestRail(section = TestGroup.SEARCH, executionType = ExecutionType.REGRESSION, description = "Total hit tracking with 0 should use default behaviour")
+    @Test(groups = TestGroup.SEARCH)
+    public void searchUsingTotalHitsLimitZeroShouldDefault()
+    {
+        SearchRequest request = req("cm:name:" + PREFIX + "*.txt");
+
+        // Setting trackTotalHitsLimit to 0 should behave as not setting it, counting all up to 10k
+        request.setLimits(new RestRequestLimitsModel(null, null, 0));
+        searchQueryService.expectTotalHitsFromQuery(request, userSite1, TOTAL_DOCUMENT_COUNT);
+    }
+
+    @TestRail(section = TestGroup.SEARCH, executionType = ExecutionType.REGRESSION, description = "Total hit tracking set should use the defined limit")
+    @Test(groups = TestGroup.SEARCH)
+    public void searchUsingTotalHitsLimit()
+    {
+        SearchRequest request = req("cm:name:" + PREFIX + "*.txt");
+
+        // Setting trackTotalHitsLimit to 10 should only count up to 10
+        request.setLimits(new RestRequestLimitsModel(null, null, 10));
+        searchQueryService.expectTotalHitsFromQuery(request, userSite1, 10);
+    }
+
+    @TestRail(section = TestGroup.SEARCH, executionType = ExecutionType.REGRESSION, description = "Total hit tracking with -1 should count all docs")
+    @Test(groups = TestGroup.SEARCH)
+    public void searchUsingTotalHitsUnlimited()
+    {
+        SearchRequest request = req("cm:name:" + PREFIX + "*.txt");
+
+        // Setting trackTotalHitsLimit to -1 should only count up to max int
```

Comment:
"only" :joy: 

It might be worth mentioning that this is limited by the total documents.